dbms/src/Columns/ColumnString.cpp

```diff
@@ -496,14 +514,50 @@ void ColumnString::countSerializeByteSize(PaddedPODArray<size_t> & byte_size) co
                 sizeAt(i));
     }
 
-    size_t size = byte_size.size();
-    for (size_t i = 0; i < size; ++i)
-        byte_size[i] += sizeof(UInt32) + sizeAt(i);
+    if constexpr (has_collator)
+    {
+        RUNTIME_CHECK(collator);
+
+        const size_t size = byte_size.size();
+        const size_t max_bytes_one_char = collator->maxBytesForOneChar();
+        for (size_t i = 0; i < size; ++i)
+        {
+            assert(sizeAt(i) >= 1);
+            // Minus 1 because of terminating zero.
+            byte_size[i] += sizeof(UInt32) + (sizeAt(i) - 1) * max_bytes_one_char;
```

Comment:
`size * max_bytes_one_char` may waste lots of memory. For example, a utf8 character is 3 bytes, `max_bytes_one_char` is 4 bytes. 4 bytes is enough but here need 12 bytes.

Reply:
done